metabase/metabase
# Update UX of authentication methods' Group Mappings Widget

Fixes https://github.com/metabase/metabase/issues/27212

Makes the "Group Mappings Widget" inlined in 3 authentication method forms inside our Admin.

Adds ability to delete groups, or empty groups of users, along with deleting a mapping.

Updates UI details.

Specifications:
[Product Doc](https://www.notion.so/metabase/Make-SSO-group-mapping-clearer-b9ed33e4ee8c4facb2220f20b565e51f)

Here we also extract components away from the `GroupMappingsWidget` file, which was large.
We apply TypeScript to extracted files.
We add test coverage to the feature using jest.

### To verify

Use application in Enterprise mode.

The widget appears in these 3 pages:

http://localhost:3000/admin/settings/authentication/ldap
http://localhost:3000/admin/settings/authentication/saml
http://localhost:3000/admin/settings/authentication/jwt

Scroll to the bottom, as the widget lives near the end of the form.

Once there, you should see it:

<img width="400" alt="image" src="https://user-images.githubusercontent.com/380816/218839251-e89304c9-0aa2-4c30-ad99-678f480e0fda.png">

You should be able to create a `New Mapping`, by giving it a unique name and hitting `Save`.
Once created, you can assign one or more groups to it.

Creating and editing a mapping happens async, that is, the changes are saved without submitting the form.

#### Deleting a mapping.

Once you've created a mapping, the delete flow can be slightly different depending on what groups if any are mapped:

**No group, or only Admin group:** a confirm modal will be shown. Clicking on `Yes` deletes the mapping.

**One or more groups besides Admin group:** a different modal will be shown. 
This will offer the possibility of:

1. Just delete the mapping. Groups will stay as they are.
2. Clear groups of their members. Groups will not be deleted, but all users in it will be removed from group.
3. Delete groups. Groups will be deleted. Users will *not* be removed from system.

To verify updates to groups after choosing option 2 or 3 above, visit http://localhost:3000/admin/people/groups.

#### Submitting the form — LDAP example

<details>
  <summary>To configure LDAP locally</summary>

[This PR](https://github.com/metabase/metabase/pull/25439) has good instructions.  

In short, you run a docker image: `docker run -p 389:389 osixia/openldap:1.5.0`

Then visit http://localhost:3000/admin/settings/authentication/ldap

Fill inputs with:

**LDAP Host:** localhost
**LDAP Port**: 389
**Username or DN:** cn=admin,dc=example,dc=org
**Password:** admin
**User search base:** dc=example,dc=org
</details>
